algorithms.py

```diff
@@ -521,6 +684,113 @@ def _make_segment(self, left, right, count):
         return seg
 
 
+class Hull:
+    def __init__(self, index):
+        self.left = None
+        self.right = None
+        self.ancestor_node = None
+        self.index = index
+        self.insertion_order = math.inf
+
+    def __lt__(self, other):
+        return (self.left, self.insertion_order) < (other.left, other.insertion_order)
+
+    def __repr__(self):
+        return f"l:{self.left}, r:{self.right}, io:{self.insertion_order}"
+
+
+class HullEnd:
+    def __init__(self, x):
+        self.x = x
+        self.insertion_order = math.inf
+
+    def __lt__(self, other):
+        return (self.x, self.insertion_order) < (other.x, other.insertion_order)
+
+    def __repr__(self):
+        return f"x:{self.x}, io:{self.insertion_order}"
+
+
+class OrderStatisticsTree:
```

Comment:
Explain here that bintrees doesn't quite do what we want and we're emulating that inefficiently.